packages/vats/src/core/chain-behaviors.js

```diff
@@ -282,6 +282,76 @@ export const makeBridgeManager = async ({
 };
 harden(makeBridgeManager);
 
+// TODO: Refine typing... maybe something like PromiseSpace or PromiseSpaceOf?
+/**
+ * @param { BootstrapPowers } powers
+ */
+export const makeChainStorage = async ({
+  consume: { bridgeManager: bridgeManagerP },
+  produce: { chainStorage: chainStorageP },
+}) => {
+  const bridgeManager = await bridgeManagerP;
+  if (!bridgeManager) {
+    console.warn('Cannot support chainStorage without an actual chain.');
+    chainStorageP.resolve(undefined);
+    return;
+  }
+  const toStorage = message =>
+    E(bridgeManager).toBridge(BRIDGE_ID.STORAGE, message);
+
+  // TODO: Formalize root key.
+  // Must not be any of {activityhash,beansOwing,egress,mailbox},
+  // and must be reserved in sites that use those keys (both Go and JS).
+  const ROOT_KEY = 'published';
+  // TODO: Formalize segment constraints.
+  // Must be nonempty and disallow (unescaped) `.`, and for simplicity
+  // (and future possibility of e.g. escaping) we currently limit to
+  // ASCII alphanumeric plus underscore.
+  const pathSegmentPattern = /^[a-zA-Z0-9_]{1,100}$/;
+  const makeChainStorageNode = key => {
+    const node = {
+      getKey() {
+        return key;
+      },
+      getChildNode(name) {
+        assert.typeof(name, 'string');
+        if (!pathSegmentPattern.test(name)) {
+          assert.fail(
+            X`Path segment must be a short ASCII identifier: ${name}`,
+          );
+        }
+        return makeChainStorageNode(`${key}.${name}`);
+      },
+      setValue(value) {
+        assert.typeof(value, 'string');
+        // TODO: Fix on the Go side.
+        // https://github.com/Agoric/agoric-sdk/issues/5381
+        assert(value !== '');
+        toStorage({ key, method: 'set', value });
+      },
+      async delete() {
+        assert(key !== ROOT_KEY);
+        const childKeys = await toStorage({ key, method: 'keys' });
+        if (childKeys.length > 0) {
```

Comment:
This is racy.  Is there a specific reason to refuse to delete nodes with children?  If so, we should fix this in Go as well.

Reply:
https://github.com/Agoric/agoric-sdk/issues/5381

Reply:
> #5381 describes the root problem and anticipates a future fix in Go.

#5394 solves this problem, conclusively.  We'll see whose PR lands first, and the subsequent one can remove the extra `await toStorage(...)`.

Reply:
I think #5394 is the right place to do this check, let's remove it from here to avoid having two different kinds of errors